### PR TITLE
Remix by shaving peaks : adapt for multi storage [ANT-3175]

### PR DESCRIPTION
**Pull request by guilpier-code**
This PR addresses [ticket ANT-3175](https://gopro-tickets.rte-france.com/browse/ANT-3175).
It follows [PR 2959](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2959) on hydro remix simplification.

What was done : 
- [x] sources about remix storage were moved from **src/solver/simulation** to **src/solver/simulation/remix-storage**
- [x] As argument of algorithm, the object associated to one storage was replaced with a `std::vector` of storage remix objects. 
- [x] what else ?

### CODE REVIEW

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    std::set<unsigned> validHours = ValidHours(Spillage, DTG_MRG);
+
+    unsigned nbLoops = 0;
+    auto storage = listStorage.begin();
+    while (!listStorage.empty() || nbLoops == maxNbLoops)
+    {
+        if (nbLoops++ == maxNbLoops)
+        {
+            throw std::runtime_error("storage remix > max nb of iterations was reached");
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp:58]: Warning here ?

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                       std::shared_ptr<IStorageForRemix> storage)
+{
+    double maxVariation = std::max(TotalGen[hourOfMaxGen] - TotalGen[hourOfMinGen], 0.);
+    double maxExchangeFromStorage = storage->maxExchange(hourOfMaxGen, hourOfMinGen);
+    return std::max(std::min(maxExchangeFromStorage, maxVariation / 2.), 0.);
+}
+
+struct Exchange
+{
+    bool valid()
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]:
```suggestion
    bool isValid() const
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    if (!Load.size())
+    {
+        throw std::invalid_argument(error_msg_start + "all arrays of sizes 0");
+    }
+}
+
+double computeExchange(unsigned hourOfMinGen,
+                       unsigned hourOfMaxGen,
+                       const std::vector<double>& TotalGen,
+                       std::shared_ptr<IStorageForRemix> storage)
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]:
```suggestion
double computeExchange(unsigned hourOfMinGen,
                       unsigned hourOfMaxGen,
                       const std::vector<double>& TotalGen,
                       IStorageForRemix& storage)
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    TotalGen[hourOfMinGen] += amount;
+}
+
+auto removeStorageFromList(std::vector<std::shared_ptr<IStorageForRemix>>::iterator storage,
+                           ListStorageForRemix& listStorage)
+{
+    auto d = std::distance(listStorage.begin(), storage);
+    listStorage.erase(storage, storage + 1);
+    return listStorage.begin() + d;
+}
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: What does this function do ? I suppose it erases an iterator, and returns the next element. What about if the deleted element was the last ?

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            }
+            validHoursForMax.erase(hourOfMaxGen);
+        }
+        validHoursForMin.erase(hourOfMinGen);
+    }
+    return {};
+}
+
+void update(Exchange& exchange,
+            std::shared_ptr<IStorageForRemix> storage,
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]:
```suggestion
void update(Exchange& exchange,
            IStorageForRemix& storage,
```

**Review by flomnes — COMMENTED** [review]
- Rename method `update()` -> `updateLevels`
> What if it's an reservoir hydro with no levels ?
- Document & test function `removeStorageFromList`
> Now a function inside the cyclic iterator, which is unit tested, and so documented as well
- `f(const double)` -> `f(double)`
- `f(const std::vector<double>)` -> `f(std::vector<double>)`
- `f(const std::shared_ptr<T>)` -> `f(T&)` in most cases

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp (hunk 4 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: > What does this function do ? I suppose it erases an iterator, and returns the next element.

it removes an element from a list (here a std::vector) and returns an iterator pointing to the element that replaces the removed one.

> What about if the deleted element was the last ?

In this case, the returned iterator points at the end of the list.

Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: Simpler and works as well : 
```c++
auto removeStorageFromList(std::vector<std::shared_ptr<IStorageForRemix>>::iterator storage,
                           ListStorageForRemix& listStorage)
{
    return listStorage.erase(storage, storage + 1);
}
```
A cyclic iterator was implemented, and unit tested. This code is part of it.

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp (hunk 2 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: Done

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    if (it_ == v_.end())
+    {
+        it_ = v_.begin();
+    }
+}
+
+template<typename T>
+CyclicIterator<T>& CyclicIterator<T>::delete_current()
+{
+    it_ = v_.erase(it_, it_ + 1);
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]:
```suggestion
    it_ = v_.erase(it_);
```

**Review by flomnes — COMMENTED** [review]
Add test for `CyclicIterator` with zero elements

> done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp (hunk 3 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: Done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp (hunk 5 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp]: Done

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h (hunk 6 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]: Done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp (hunk 1 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp:58]: The algorithm is very light and easy to compile as it is.
If we decide to log here, we would have to link against the logger, it would make the algorithm lib heavier and longer to compile, as we would have to make sure logger is compiled first before compiling it.
On the other hand, exceptions thrown by this remix algorithm are caught by the caller, which logs the exceptions messages.

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+template<typename T>
+class CyclicIterator
+{
+public:
+    CyclicIterator(std::vector<T>& v);
+    CyclicIterator& operator++(int);
+    T& operator*();
+    bool operator==(const std::vector<T>::iterator& it);
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]:
```suggestion
    bool operator==(const std::vector<T>::iterator& other) const;
```

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+T& CyclicIterator<T>::operator*()
+{
+    return *it_;
+}
+
+template<typename T>
+bool CyclicIterator<T>::operator==(const typename std::vector<T>::iterator& it)
+{
+    return it_ == it;
+}
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]:
```suggestion
template<typename T>
bool CyclicIterator<T>::operator==(const typename std::vector<T>::iterator& other) const
{
    return it_ == other;
}
```

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                              const std::vector<double> Pmax,
+                              const std::vector<double> Pmin,
+                              const std::vector<double> inflows,
+                              const std::vector<double> overflow,
+                              const std::vector<double>& injection,
+                              const std::vector<double> lowRuleCurve,
+                              const std::vector<double> upRuleCurve,
+                              const double initLevel,
+                              const double withdrawalEff,
+                              const double injectionEff);
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h]: I see that the caller must provide a `withdrawalEff` & `injectionEff` at construction, while there are also
```cpp
    const double withdrawalEff_ = 1;
    const double injectionEff_ = 1;
```
right below. Why ?

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                                                    const std::vector<double>& Pmin,
+                                                    const std::vector<double>& inflows,
+                                                    const std::vector<double>& overflow,
+                                                    const std::vector<double>& pump,
+                                                    const double initLevel,
+                                                    const double reservoirCapacity,
+                                                    const double pumpEfficiency,
+                                                    bool reservoirManagement);
+
+std::shared_ptr<StorageForRemixWithLevels> makeSTSforRemix(std::vector<double>& withdrawal,
```
Comment (flomnes) [inline src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h]:
```suggestion
std::shared_ptr<IStorageForRemix> makeSTSforRemix(std::vector<double>& withdrawal,
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+#include <ranges>
+
+namespace vws = std::views;
+
+constexpr double eps = 1e-3;
+
+namespace Antares::Solver::Simulation
+{
+
+bool Exchange::isPossible()
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]:
```suggestion
bool Exchange::isPossible() const
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    return amount > eps && hourOfMinGen.has_value() && hourOfMaxGen.has_value();
+}
+
+std::set<unsigned> ValidHours(const std::vector<double>& Spillage,
+                              const std::vector<double>& DTG_MRG)
+{
+    auto filter = [&](int h) { return std::abs(Spillage[h] + DTG_MRG[h]) < eps; };
+    auto validHoursView = vws::iota(0, (int)Spillage.size()) | vws::filter(filter);
+    return {validHoursView.begin(), validHoursView.end()};
+}
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]:
```suggestion

std::set<unsigned> ValidHours(const std::vector<double>& spillage,
                              const std::vector<double>& dtg_mrg)
{
    auto isValidHour = [&](int h) {
        return std::abs(spillage[h] + dtg_mrg[h]) < eps;
    };

    auto hours = std::views::iota(0, static_cast<int>(spillage.size()));
    auto validHours = hours | std::views::filter(isValidHour);

    return {validHours.begin(), validHours.end()};
}
}
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+void updateValidHours(std::set<unsigned>& validHours,
+                      std::shared_ptr<IStorageForRemix> storage,
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]:
```suggestion
                      IStorageForRemix& storage,
```

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+                return {*hourOfMinGen, *hourOfMaxGen, exchange};
+            }
+            validHoursForMax.erase(hourOfMaxGen);
+        }
+        validHoursForMin.erase(hourOfMinGen);
+    }
+    return {};
+}
+
+void update(Exchange& exchange,
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]:
```suggestion
void update(const Exchange& exchange,
```

File: src/solver/simulation/remix-storage/storage-for-remix.cpp [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+                                                       overflow,
+                                                       pump,
+                                                       lowRuleCurve,
+                                                       upRuleCurve,
+                                                       initLevel,
+                                                       withdrawalEff,
+                                                       pumpEfficiency);
+}
+
+std::shared_ptr<StorageForRemixWithLevels> makeSTSforRemix(std::vector<double>& withdrawal,
```
Comment (flomnes) [inline src/solver/simulation/remix-storage/storage-for-remix.cpp]:
```suggestion
std::shared_ptr<IStorageForRemix> makeSTSforRemix(std::vector<double>& withdrawal,
```

File: src/solver/simulation/remix-storage/storage-for-remix.cpp
Comment (flomnes) [inline src/solver/simulation/remix-storage/storage-for-remix.cpp:1]: Please split this file (`StorageForRemixNoLevels`, `StorageForRemixWithLevels`, factory method, etc.)

File: src/tests/src/solver/simulation/remix-storage/tests-on-remix-utils.cpp [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    std::string err_msg = "call of min_on_subrange : hour out of bound";
+    BOOST_CHECK_EXCEPTION(min_on_subrange({1, 2}, -1, 2),
+                          std::invalid_argument,
+                          checkMessage(err_msg));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
+
+BOOST_AUTO_TEST_SUITE(cyclic_iterator_unit_tests)
+
```
Comment (flomnes) [inline src/tests/src/solver/simulation/remix-storage/tests-on-remix-utils.cpp:162]: Add tests on `operator==(std::vector<int>::iterator)`

File: src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+#include <vector>
+
+#include <boost/test/unit_test.hpp>
+
+#include "antares/solver/simulation/remix-storage/shave-peaks-by-remix-help.h"
+#include "antares/solver/simulation/remix-storage/storage-for-remix.h"
+
+using namespace Antares::Solver::Simulation;
+
+constexpr double infinity = 1000;
```
Comment (flomnes) [inline src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp]:
```suggestion

constexpr double largeValue = 1000;
```

File: src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+    validHours = {1, 3, 6};
+    auto storage = createSTSstorage();
+
+    auto exchange = searchForExhange(validHours, totalGen, unsupE, *storage);
+
+    BOOST_CHECK(exchange.hourOfMinGen.has_value());
+    BOOST_CHECK(exchange.hourOfMaxGen.has_value());
+
+    BOOST_CHECK_EQUAL(exchange.hourOfMaxGen.value(), 1);
+    BOOST_CHECK_EQUAL(exchange.hourOfMinGen.value(), 6);
```
Comment (flomnes) [inline src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp]:
```suggestion
    BOOST_CHECK_EQUAL(exchange.hourOfMaxGen.value_or(0), 1);
    BOOST_CHECK_EQUAL(exchange.hourOfMinGen.value_or(0), 6);
```

**Review by flomnes — COMMENTED** [review]
Minor remarks, style & naming

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp (hunk 12 above)
Comment (flomnes) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]: The suggested code looks more readable in my opinion

File: src/solver/simulation/remix-storage/storage-for-remix.cpp [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+
+std::vector<double>& StorageForRemixNoLevels::withdrawal()
+{
+    return withdrawal_;
+}
+
+StorageForRemixWithLevels::StorageForRemixWithLevels(std::vector<double>& withdrawal,
+                                                     std::vector<double>& unsupE,
+                                                     std::vector<double>& levels,
+                                                     const std::vector<double> Pmax,
```
Comment (guilpier-code) [inline src/solver/simulation/remix-storage/storage-for-remix.cpp]: if you take the example of **Pmax**, it was just copied when calling the current constructor.
Then we forward it to constructor of **StorageForRemixNoLevels** by a copy, and then we store it by making another copy.

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h (hunk 7 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]: Done

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h (hunk 8 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/remix-storage/remix-utils.h]: Done

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h (hunk 9 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h]: Constructor give these 2 a default value now

File: src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h (hunk 10 above)
Comment (guilpier-code) [reply src/solver/simulation/include/antares/solver/simulation/remix-storage/storage-for-remix.h]: Done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp (hunk 11 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]: Done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp (hunk 12 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]: Right. Done.

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp (hunk 13 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]: Done

File: src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp (hunk 14 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/shave-peaks-by-remix-help.cpp]: Done

File: src/solver/simulation/remix-storage/storage-for-remix.cpp (hunk 15 above)
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/storage-for-remix.cpp]: Done

File: src/solver/simulation/remix-storage/storage-for-remix.cpp
Comment (guilpier-code) [reply src/solver/simulation/remix-storage/storage-for-remix.cpp:1]: Done

File: src/tests/src/solver/simulation/remix-storage/tests-on-remix-utils.cpp (hunk 16 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/remix-storage/tests-on-remix-utils.cpp:162]: Added 2 unit tests

File: src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp (hunk 17 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp]: Done

File: src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp (hunk 18 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/remix-storage/tests-on-shave-peak-by-remix-help.cpp]: Shorter indeed. Done everywhere I could.

**Review by flomnes — APPROVED** [review]
